core/state/state_object.go

```diff
@@ -395,6 +395,43 @@ func (s *stateObject) CommitTrie(db Database) (*trie.NodeSet, error) {
 	return nodes, err
 }
 
+// DeleteTrie the storage trie of the object from db.
+func (s *stateObject) DeleteTrie(db Database) (*trie.NodeSet, error) {
+	// Track the amount of time wasted on iterating and deleting the storage trie
+	if metrics.EnabledExpensive {
+		defer func(start time.Time) { s.db.StorageDeletes += time.Since(start) }(time.Now())
+	}
+	stTrie, err := db.OpenStorageTrie(s.db.originalRoot, s.addrHash, s.data.Root)
+	if err != nil {
+		return nil, err
+	}
+	// It can be an attack vector when iterating a huge contract. Stop collecting
+	// in case the accumulated nodes reach the threshold. It's fine to not clean
+	// up the dangling trie nodes since they are non-accessible dangling nodes
+	// anyway.
+	var (
+		paths [][]byte
+		blobs [][]byte
+		size  common.StorageSize
+		iter  = stTrie.NodeIterator(nil)
+	)
+	for iter.Next(true) {
+		if iter.Hash() == (common.Hash{}) {
+			continue
+		}
+		path, blob := common.CopyBytes(iter.Path()), common.CopyBytes(iter.NodeBlob())
```

Comment:
What is the need for `prev` again, when storing away the to-be-deleted nodes? 
In other words: why do we hold on to the values?

Reply:
the `prev` value of `deleted nodes` is also for building the reverse diff. When we apply the reverse diff, we just persist the prev value into database.